### PR TITLE
Implement DataProxy CreateUploadLocation API

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
Implement DataProxy's CreateUploadLocation API

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue

https://github.com/flyteorg/flyte/issues/2263

### CODE REVIEW

File: dataproxy/service.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/flyteorg/flytestdlib/storage"
+	"github.com/flyteorg/stow"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+)
+
+type Service struct {
+	service.DataProxyServer
+	*service.UnimplementedDataProxyServer
```
Comment (katrogan) [inline dataproxy/service.go:26]: do we want this?

File: dataproxy/service.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		if expiresIn.AsDuration() > s.cfg.Upload.MaxExpiresIn.Duration {
+			return nil, fmt.Errorf("expiresIn [%v] cannot exceed max allowed expiration [%v]",
+				expiresIn.AsDuration().String(), s.cfg.Upload.MaxExpiresIn.String())
+		}
+	} else {
+		req.ExpiresIn = durationpb.New(s.cfg.Upload.MaxExpiresIn.Duration)
+	}
+
+	if len(req.Suffix) == 0 {
+		req.Suffix = rand.String(20)
```
Comment (katrogan) [inline dataproxy/service.go:55]: why is this necessary?

File: dataproxy/service.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	return &service.CreateUploadLocationResponse{
+		SignedUrl: resp.URL.String(),
+		NativeUrl: storagePath.String(),
+		ExpiresAt: timestamppb.New(time.Now().Add(req.ExpiresIn.AsDuration())),
+	}, nil
+}
+
+func createShardedStorageLocation(ctx context.Context, req *service.CreateUploadLocationRequest,
```
Comment (katrogan) [inline dataproxy/service.go:80]: interesting, does this logic not currently exist elsewhere?

File: pkg/config/config.go [hunk 4]
```diff
@@ -64,6 +80,13 @@ var defaultServerConfig = &ServerConfig{
 		Port:             8089,
 		ServerReflection: true,
 	},
+	DataProxy: DataProxyConfig{
+		Upload: DataProxyUploadConfig{
+			MaxSize:               resource.MustParse("6Mi"),
+			MaxExpiresIn:          config.Duration{Duration: time.Hour},
```
Comment (katrogan) [inline pkg/config/config.go:86]: this is for fast registration, right? why so short by default?

File: pkg/rpc/adminservice/base.go [hunk 5]
```diff
@@ -5,6 +5,10 @@ import (
 	"fmt"
 	"runtime/debug"
 
+	"github.com/flyteorg/flyteadmin/plugins"
+
+	interfaces2 "github.com/flyteorg/flyteadmin/pkg/runtime/interfaces"
```
Comment (katrogan) [inline pkg/rpc/adminservice/base.go:10]: nit: interfaces2

File: pkg/rpc/adminservice/base.go [hunk 6]
```diff
@@ -93,13 +93,7 @@ func NewAdminServer(ctx context.Context, kubeConfig, master string) *AdminServic
...
 	workflowExecutor := workflowengineImpl.NewK8sWorkflowExecutor(execCluster, workflowBuilder)
 	logger.Info(ctx, "Successfully created a workflow executor engine")
-	workflowengine.GetRegistry().RegisterDefault(workflowExecutor)
-
-	dataStorageClient, err := storage.NewDataStore(storeConfig, adminScope.NewSubScope("storage"))
-	if err != nil {
-		logger.Error(ctx, "Failed to initialize storage config")
-		panic(err)
-	}
+	pluginRegistry.RegisterDefault(plugins.PluginIDWorkflowExecutor, workflowExecutor)
```
Comment (katrogan) [inline pkg/rpc/adminservice/base.go:96]: nice!

**Review by katrogan — APPROVED** [review]
thanks for adding the plugin registry

File: dataproxy/service.go (hunk 3 above)
Comment (EngHabu) [reply dataproxy/service.go:80]: This is the closest to this logic: https://github.com/flyteorg/flyteplugins/blob/60b94c688ef2b98aa53a9224b529ac672af04540/go/tasks/pluginmachinery/ioutils/raw_output_path.go#L55-L68

File: pkg/config/config.go (hunk 4 above)
Comment (EngHabu) [reply pkg/config/config.go:86]: It's indeed for fast registration. Is an hour not enough to upload for fast register? How long do you think it should be?
